src/Workspaces/Remote/ServiceHub/Host/RemoteSolutionCache.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+/// <summary>
+/// LRU cache of checksum+solution pairs.  Used to keep track of the last few solutions the remote server knows about,
+/// helping to avoid unnecessary syncs/recreations of those solutions while many requests are coming into the server.
+/// Not threadsafe.  Only use while under a lock.
+/// </summary>
+internal sealed class RemoteSolutionCache<TChecksum, TSolution>
+    where TChecksum : struct
+    where TSolution : class
+{
+    /// <summary>
+    /// The max number of solution instances we'll hold onto at a time.
+    /// </summary>
+    private readonly int _maxCapacity;
+
+    /// <summary>
+    /// The total history kept.  Used to record telemetry about how useful it would be to increase the max capacity.
+    /// </summary>
+    private readonly int _totalHistory;
+
+    /// <summary>
+    /// Keep track of when we find a checksum in the history, but we've dropped the solution for it.  This will help
+    /// us determine what benefit we would get from expanding this cache.
+    /// </summary>
+    private readonly HistogramLogAggregator<int>.HistogramCounter _cacheMissCounter;
```

Comment:
>_cacheMissCounter [](http://example.com/codeflow?start=66&length=17)

What do you think about just keeping all request indexes instead of just the misses? That would tell us if we are caching too many items too

Reply:
sure.  i'll change to that.